### ISSUE TITLE
Some dependencies in geonode/static/lib/js are tracked in git while others are ignored

### ISSUE THREAD

**Issue by harts-boundless** [GeoNode/geonode#2669]
The way that dependencies are treated is inconsistent, with some files treated one way and other files treated another way, and the choice seems to be arbitrary.

When you get a GeoNode clone, you get certain frontend dependencies. (Whether this is optimal or not is perhaps another topic.) But you don't get certain other frontend dependencies and you don't know they're missing until you see 404s in the browser. This is not a great way of discovering it, and it doesn't make it very obvious what is wrong and how to fix it.

cleanup proposals
--------

1. remove references to frontend dependencies which are not brought in by `npm install --save-dev` and are not used (see "Listed in gitignore")

2. remove other references to frontend dependencies which are not referenced anywhere but gruntfile.js (mentioned throughout the below sections) and test carefully.

3. remove listings from .gitignore which point to files which are already tracked as part of the git repo (see "Listed in gitignore", the first part of it).

4. remove listings from .gitignore which point to essential files (basically remove all the .gitignore listings from "Listed in gitignore").

5. make the way that frontend files are tracked in git consistent (choose one of the following two approaches):

a. add all used frontend files to the repo for consistency (see "not in git"). any further cleanups could be done later. While GeoNode is a python package, this would follow the npm convention of providing built files in the repo for direct use in a checkout instead of requiring a frontend build. It would also be least likely to break any existing GeoNode projects downstream which are unaware of these changes. It is just a bit untidy.

b. ensure checked-in versions are exactly specified in `bower.json`, then remove all the corresponding files from the git repo. make sure any automation scripts in the repo pull these down where collectstatic will get at them, and that documentation like README mentions the need to do this in any GeoNode deployment or GeoNode project. This seems tidier for development but as a user or downstream project its acceptability depends on how likely I was to have the frontend build not done.

in git
------

These are relative paths of files found in git ls-files.

Underneath each filename is a list of code files which appear to reference it

```
+css/assets.min.css
    catalogue/templates/geonode_metadata_full.html
    static/gruntfile.js
    templates/500.html
    templates/base.html
+css/bootstrap-datetimepicker.css
    static/gruntfile.js
    templates/500.html
    templates/base.html
    templates/geonode_metadata_full.html
+css/L.Control.Pan.css
    maps/templates/leaflet_maps/map_view.html
+css/leaflet.fullscreen.css
    settings.py (in LEAFLET_SETTINGS.PLUGINS)
    static/gruntfile.js
+js/assets.min.js
    static/gruntfile.js
    templates/500.html
    templates/admin/bsae_site.html
    templates/base.html
+js/bootstrap-datetimepicker.min.js
    static/gruntfile.js
    templates/base.html
    templates/metadata_form_js.html
+js/bootstrap.js
    js/tests/run-tests.sh
    static/gruntfile.js
+js/esri-leaflet.js
    settings.py (in LEAFLET_SETTINGS.PLUGINS)
+js/L.Control.Pan.js
    maps/templates/leaflet_maps/map_view.html
+js/Leaflet.fullscreen.min.js
    settings.py (in LEAFLET_SETTINGS.PLUGINS) 
    static/gruntfile.js
+js/moment-with-locales.min.js
    static/gruntfile.js
    templates/base.html
+js/require.js
    layers/templates/layers/layer_replace.html
    layers/templates/upload/layer_upload_crs.html
    layers/templates/upload/layer_upload_csv.html
    layers/templates/upload/layer_upload.html
    layers/templates/upload/layer_upload_time.html
    static/gruntfile.js
+js/text.js
    static/gruntfile.js
+js/underscore.js
    static/gruntfile.js
```

text.js and underscore.js may not have any consumers in GeoNode itself.

not in git
----------

These are files present only after `npm install --save-dev`, but untracked by
git. However, git doesn't show these in git status, because they are all listed
in .gitignore. So if they changed or are completely missing, `git status`
cannot tell you and you will not know until you see 404s in the browser.

```
-css/bootstrap-tokenfield.css
    static/gruntfile.js
    templates/base.html
-css/bootstrap-treeview.min.css
    static/gruntfile.js
-css/jquery.dataTables.css
    catalogue/templates/geonode_metadata_full.html
    static/gruntfile.js
    templates/500.html
    templates/base.html
-css/jquery-ui.css
    catalogue/templates/geonode_metadata_full.html
    maps/templates/leaflet_maps/map_view.html
    static/gruntfile.js
    templates/500.html
    templates/base.html
-css/multi-select.css
    static/gruntfile.js
-css/select2.css
    catalogue/templates/geonode_metadata_full.html
    static/gruntfile.js
    templates/500.html
    templates/base.html
-js/bootstrap-tokenfield.js
    static/gruntfile.js    
    templates/metadata_form_js.html
-js/bootstrap-treeview.min.js
    static/gruntfile.js
    templates/metadata_form_js.html
    templates/search/search_scripts.html
-js/jquery.ajaxprogress.js
    static/gruntfile.js
    templates/500.html
    templates/base.html
-js/jquery.ajaxQueue.js
    static/gruntfile.js
    templates/500.html
    templates/base.html
-js/jquery.dataTables.js
    static/gruntfile.js
    templates/500.html
    templates/base.html
-js/jquery.multi-select.js
    static/gruntfile.js
-js/jquery.timeago.js
    static/gruntfile.js
    templates/500.html
    templates/base.html
-js/jquery-ui.custom.js
    static/gruntfile.js
    templates/500.html
    templates/base.html
-js/json2.js
    static/gruntfile.js
    templates/500.html
    templates/base.html
-js/select2.js
    static/gruntfile.js
    templates/500.html
    templates/base.html
-js/waypoints.js
    static/gruntfile.js
    templates/500.html
    templates/base.html
-js/ZeroClipboard.min.js
    contrib/geogig/templates/_geogig_scripts.html
    static/gruntfile.js
```

bootstrap-treeview.min.css and multi-select.css and jquery.multi-select.js may
not be used anywhere.

Listed in .gitignore
--------------------

(but already checked into git, so they are tracked anyway)

```
+geonode/static/lib/css/assets.min.css
+geonode/static/lib/js/assets.min.js
+geonode/static/lib/js/bootstrap.js
+geonode/static/lib/js/require.js
+geonode/static/lib/js/text.js
+geonode/static/lib/js/underscore.js
```

(not in git because listed in .gitignore)
```
-geonode/static/lib/css/bootstrap-tokenfield.css
-geonode/static/lib/css/bootstrap-treeview.min.css
-geonode/static/lib/css/jquery.dataTables.css
-geonode/static/lib/css/jquery-ui.css
-geonode/static/lib/css/multi-select.css
-geonode/static/lib/css/select2.css
-geonode/static/lib/js/bootstrap-tokenfield.js
-geonode/static/lib/js/bootstrap-treeview.min.js
-geonode/static/lib/js/jquery.ajaxprogress.js
-geonode/static/lib/js/jquery.ajaxQueue.js
-geonode/static/lib/js/jquery.dataTables.js
-geonode/static/lib/js/jquery.multi-select.js
-geonode/static/lib/js/jquery.timeago.js
-geonode/static/lib/js/jquery.tinysort.js
-geonode/static/lib/js/jquery-ui.custom.js
-geonode/static/lib/js/json2.js
-geonode/static/lib/js/select2.js
-geonode/static/lib/js/waypoints.js
-geonode/static/lib/js/ZeroClipboard.min.js
```

(listed in .gitignore but not actually present after npm install --save-dev)
```
?geonode/static/lib/css/bootstrap-treeview.css
?geonode/static/lib/css/qunit.css
?geonode/static/lib/js/angular.min.js
?geonode/static/lib/js/angular.min.js.map
?geonode/static/lib/js/bootstrap-datepicker.js
?geonode/static/lib/js/bootstrap-treeview.js
?geonode/static/lib/js/jquery.js
?geonode/static/lib/js/jquery.raty.js
?geonode/static/lib/js/moment.min.js
?geonode/static/lib/js/qunit.js
```

with the exception of jquery.raty.js and qunit.js, none of the files in this
last group seem to be used in GeoNode.

in git
------

These are relative paths of files found in git ls-files.

Underneath each filename is a list of code files which appear to reference it

```
+css/assets.min.css
    catalogue/templates/geonode_metadata_full.html
    static/gruntfile.js
    templates/500.html
    templates/base.html
+css/bootstrap-datetimepicker.css
    static/gruntfile.js
    templates/500.html
    templates/base.html
    templates/geonode_metadata_full.html
+css/L.Control.Pan.css
    maps/templates/leaflet_maps/map_view.html
+css/leaflet.fullscreen.css
    settings.py (in LEAFLET_SETTINGS.PLUGINS)
    static/gruntfile.js
+js/assets.min.js
    static/gruntfile.js
    templates/500.html
    templates/admin/bsae_site.html
    templates/base.html
+js/bootstrap-datetimepicker.min.js
    static/gruntfile.js
    templates/base.html
    templates/metadata_form_js.html
+js/bootstrap.js
    js/tests/run-tests.sh
    static/gruntfile.js
+js/esri-leaflet.js
    settings.py (in LEAFLET_SETTINGS.PLUGINS)
+js/L.Control.Pan.js
    maps/templates/leaflet_maps/map_view.html
+js/Leaflet.fullscreen.min.js
    settings.py (in LEAFLET_SETTINGS.PLUGINS) 
    static/gruntfile.js
+js/moment-with-locales.min.js
    static/gruntfile.js
    templates/base.html
+js/require.js
    layers/templates/layers/layer_replace.html
    layers/templates/upload/layer_upload_crs.html
    layers/templates/upload/layer_upload_csv.html
    layers/templates/upload/layer_upload.html
    layers/templates/upload/layer_upload_time.html
    static/gruntfile.js
+js/text.js
    static/gruntfile.js
+js/underscore.js
    static/gruntfile.js
```

text.js and underscore.js may not have any consumers in GeoNode itself.

not in git
----------

These are files present only after `npm install --save-dev`, but untracked by
git. However, git doesn't show these in git status, because they are all listed
in .gitignore. So if they changed or are completely missing, `git status`
cannot tell you and you will not know until you see 404s in the browser.

```
-css/bootstrap-tokenfield.css
    static/gruntfile.js
    templates/base.html
-css/bootstrap-treeview.min.css
    static/gruntfile.js
-css/jquery.dataTables.css
    catalogue/templates/geonode_metadata_full.html
    static/gruntfile.js
    templates/500.html
    templates/base.html
-css/jquery-ui.css
    catalogue/templates/geonode_metadata_full.html
    maps/templates/leaflet_maps/map_view.html
    static/gruntfile.js
    templates/500.html
    templates/base.html
-css/multi-select.css
    static/gruntfile.js
-css/select2.css
    catalogue/templates/geonode_metadata_full.html
    static/gruntfile.js
    templates/500.html
    templates/base.html
-js/bootstrap-tokenfield.js
    static/gruntfile.js    
    templates/metadata_form_js.html
-js/bootstrap-treeview.min.js
    static/gruntfile.js
    templates/metadata_form_js.html
    templates/search/search_scripts.html
-js/jquery.ajaxprogress.js
    static/gruntfile.js
    templates/500.html
    templates/base.html
-js/jquery.ajaxQueue.js
    static/gruntfile.js
    templates/500.html
    templates/base.html
-js/jquery.dataTables.js
    static/gruntfile.js
    templates/500.html
    templates/base.html
-js/jquery.multi-select.js
    static/gruntfile.js
-js/jquery.timeago.js
    static/gruntfile.js
    templates/500.html
    templates/base.html
-js/jquery-ui.custom.js
    static/gruntfile.js
    templates/500.html
    templates/base.html
-js/json2.js
    static/gruntfile.js
    templates/500.html
    templates/base.html
-js/select2.js
    static/gruntfile.js
    templates/500.html
    templates/base.html
-js/waypoints.js
    static/gruntfile.js
    templates/500.html
    templates/base.html
-js/ZeroClipboard.min.js
    contrib/geogig/templates/_geogig_scripts.html
    static/gruntfile.js
```

bootstrap-treeview.min.css and multi-select.css and jquery.multi-select.js may
not be used anywhere.

Listed in .gitignore
--------------------

(but already checked into git, so they are tracked anyway)

```
+geonode/static/lib/css/assets.min.css
+geonode/static/lib/js/assets.min.js
+geonode/static/lib/js/bootstrap.js
+geonode/static/lib/js/require.js
+geonode/static/lib/js/text.js
+geonode/static/lib/js/underscore.js
```

(not in git because listed in .gitignore)
```
-geonode/static/lib/css/bootstrap-tokenfield.css
-geonode/static/lib/css/bootstrap-treeview.min.css
-geonode/static/lib/css/jquery.dataTables.css
-geonode/static/lib/css/jquery-ui.css
-geonode/static/lib/css/multi-select.css
-geonode/static/lib/css/select2.css
-geonode/static/lib/js/bootstrap-tokenfield.js
-geonode/static/lib/js/bootstrap-treeview.min.js
-geonode/static/lib/js/jquery.ajaxprogress.js
-geonode/static/lib/js/jquery.ajaxQueue.js
-geonode/static/lib/js/jquery.dataTables.js
-geonode/static/lib/js/jquery.multi-select.js
-geonode/static/lib/js/jquery.timeago.js
-geonode/static/lib/js/jquery.tinysort.js
-geonode/static/lib/js/jquery-ui.custom.js
-geonode/static/lib/js/json2.js
-geonode/static/lib/js/select2.js
-geonode/static/lib/js/waypoints.js
-geonode/static/lib/js/ZeroClipboard.min.js
```

(listed in .gitignore but not actually present after `npm install --save-dev`)
```
?geonode/static/lib/css/bootstrap-treeview.css
?geonode/static/lib/css/qunit.css
?geonode/static/lib/js/angular.min.js
?geonode/static/lib/js/angular.min.js.map
?geonode/static/lib/js/bootstrap-datepicker.js
?geonode/static/lib/js/bootstrap-treeview.js
?geonode/static/lib/js/jquery.js
?geonode/static/lib/js/jquery.raty.js
?geonode/static/lib/js/moment.min.js
?geonode/static/lib/js/qunit.js
```

with the exception of jquery.raty.js and qunit.js, which are used for testing, none of the files in this last group seem to be used in GeoNode.

**Comment by harts-boundless** [comment]
Of course some of this is causally explained by certain things being dev dependencies specified in the gruntfile, but that doesn't explain why some files used by the app are gitignored and some are not, or why some gitignored files are also tracked.

**Comment by yecarrillo** [comment]
Can´t make this work! How can I add those missing files ? "python manage.py collectstatic" do not bring those files even with "DEBUG_STATIC = 'False'"
